### PR TITLE
Respect output coder for TextIO.

### PR DESCRIPTION
**Pull request by robertwb**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @pabloem

**Review by pabloem — APPROVED** [review]

**Comment by apilloud** [comment]
Looks like this may have broken some postcommits? https://issues.apache.org/jira/browse/BEAM-14351

File: sdks/python/apache_beam/io/textio.py [hunk 1]
```diff
@@ -675,7 +686,8 @@ def __init__(
         escapechar=escapechar)
 
   def expand(self, pvalue):
-    return pvalue.pipeline | Read(self._source)
+    return pvalue.pipeline | Read(self._source).with_output_types(
+        self._source.output_type_hint())
```
Comment (KevinGG) [inline sdks/python/apache_beam/io/textio.py:690]: Just a note: this is a non-backward-compatible change and will break if the user code sets `with_output_types` especially when it's an auto-generated Row schema (for Beam >= v2.39.0, the fix on user-side is: users no longer need to set `with_output_types`)

Comment (liferoad) [reply sdks/python/apache_beam/io/textio.py:690]: We should document this somewhere explicitly.

**Pull request merged**